### PR TITLE
remove fd sources from glib

### PR DESCRIPTION
**Pull request by haircommander**
or else we may close them without removing them, causing an infinte EPOLL issue

fixes https://github.com/containers/conmon/issues/167

Signed-off-by: Peter Hunt <pehunt@redhat.com>

### CODE REVIEW

File: src/ctr_exit.c [hunk 1]
```diff
@@ -151,6 +152,7 @@ void do_exit_command()
 		_pexit("Failed to reset signal for SIGCHLD");
 	}
 
+	remove_g_unix_fds();
```
Comment (cgwalters) [inline src/ctr_exit.c:155]: This does seem like a potential bug, but I'm not sure it's the actual bug here.  This problem can only occur if we reenter `g_main_loop_run()` after closing a fd behind its back.

I am not seeing where we'd being doing that.  It looks like inside `do_exit_command()` here we do some prep stuff and then just `execve()`.

Are you sure there's not another place where we're closing a fd while we're inside the `g_main_loop_run()` invocation?

**Comment by cgwalters** [comment]
Looking at this a little bit more, the logic in `stdio_cb()` seems suspect here; notice we're closing the fd inside the callback for that fd, then returning `g_source_remove()` which should remove the source.  But I'm not sure it's *guaranteed* that the main loop won't try to reenter epoll.

I think what we really want here is (in Rust terms) for the `GSource` to own the fd, not just reference it.  Then we'd only have one static array (of sources) and not two (one of sources, one of fds).

We may be able to do this with `g_unix_fd_add_full()` and pass the fd as user data, and invoke `close()` there.

**Comment by haircommander** [comment]
I have tried two different methods than the one you've described: one where we remove the source immediately after accepting on the corresponding socket https://github.com/containers/conmon/pull/313
and one here where we make sure to remove the source when we close it (here)

**Comment by haircommander** [comment]
oh you said stdio_sb, I focused on attach_cb as I found a case where we didn't remove the source. gotta check stdio_cb too...

**Comment by haircommander** [comment]
> notice we're closing the fd inside the callback for that fd, then returning g_source_remove() which should remove the source. But I'm not sure it's guaranteed that the main loop won't try to reenter epoll.

I feel like we'd be in trouble if G_SOURCE_REMOVE was or wasn't removing the fds. It's a programming error to double remove a source (as they reuse the source guints), and we do it all over the place. We use it all over the place to remove the fds. Except for in attach_cb afaict

**Comment by rhatdan** [comment]
@haircommander is this still needed since other PR fixed the original issue?

**Comment by haircommander** [comment]
nope! closing

**Pull request closed without merging**